### DISCUSSION TITLE
Adding a cuTensor backend to fl::Tensor

### DISCUSSION THREAD

**Discussion by jacobkahn** [flashlight/flashlight#749] [General]
cc @mtmd

As we wrap up the `fl::Tensor` interface and finish integrating it broadly into Flashlight, we'd like to explore integration of other NVIDIA libraries with Flashlight. cuTensor is a great candidate for a fast, NVIDIA-only backend that would feature cuTensor (along with cuBLAS, cuSparse, cuDNN, NCCL, CUDA Graphs, etc!) and would have access to the Flashlight benchmarking suite for end-to-end performance optimization. This would be a great way for NVIDIA to show off the prowess of their native CUDA libraries.

For cuTensor, this would provide a dynamic training-based end-to-end deep learning stack with minimal overhead that's based on the library. It would create a bridge for cuTensor to be used as a backend to a full-fledged research framework and provide performance context next to other tensor frameworks (e.g. ArrayFire). With an NVIDIA-only backend, Flashlight would have no dependencies outside of CUDA, cuDNN, NCCL, and related libraries, and could be ported to interesting new hardware environments.

### Design Sketch
`TensorAdapter` — representing an individual Tensor state and operations on that tensor:

```cpp
#include <cutensor.h>

class CuTensorTensor : public fl::TensorAdapter {
  // Other state of your choice
  cutensorTensorDescriptor_t descriptor_;

 public:
  CuTensorTensor(void* data, fl::dtype type) {
   HANDLE_ERROR(cutensorInitTensorDescriptor(&CuTensorBackend::getHandle(),
                &descriptor_,
                modes,
                data,
                NULL,/*stride*/
                detail::flTypeToCuTensorType(type), CUTENSOR_OP_IDENTITY));
  }

  const Shape& shape() const override;
  // ...
  Tensor flatten() const override;
  // ...
};
```

`TensorBackend` — tracking global backend state and providing operations on other tensors. 
```cpp
#include <cutensor.h>

class CuTensorBackend : public fl::TensorBackend {
  // State of your choice - global state for ops, compute streams, etc.
  cutensorHandle_t handle_;

 public:
  CuTensorBackend() {
   cutensorInit(&handle_);
   // ... -- other preflight things
  }

  cutensorHandle_t getHandle() const { return handle_; }

  fl::Tensor operator+(const fl::Tensor& lhs, const fl::Tensor& rhs) override;
  // ...
};
```

**Comment by alexjwilliams** [comment]
Are there plans to build the above mentioned backend? I would be really excited to see this, given that arrayfire is limited to 4D tensors.

**Reply by jacobkahn** [reply]
Yes — this is still in the works/planning.

The best way to integrate this would be under the JIT Tensor backend framework that @strongerxi is building (so that we can take full advantage of the fusion and composed operations in the cuTensor backend. It'll be some time until this is ready.
